Modules/_xxsubinterpretersmodule.c

```diff
@@ -347,12 +347,165 @@ channel_exceptions_init(PyObject *ns)
+{
+    assert(wait_lock != NULL);
+    assert(wait_lock->is_sent == 0);
+    PyThread_acquire_lock(wait_lock->lock, WAIT_LOCK);
+    wait_lock->is_sent = 1;
+}
+
+/**
+ * To update the wait lock's state when the channel item is removed from the
+ * channel's queue.
+ * This function must be called by the interpeter receiving the message.
+ *
+ * If received is:
+ *  -  1: then the channel item was received by an interpreter.
+ *  -  0: then the channel item was removed from the queue, or the queue was
+ *        closed, before it could be received.
+ *
+ * If this function is called after _channelitem_wait_lock_wait has timed out
+ * then it is responsible for freeing the wait lock.
+ *
+ * NOTE: This must be called after _channelitem_wait_lock_sent has been called.
+ */
+static void
+_channelitem_wait_lock_recv(_channelitem_wait_lock *wait_lock, int received)
+{
+    assert(wait_lock != NULL);
+    assert(wait_lock->is_sent == 1);
+    PyThread_type_lock mutex = wait_lock->mutex;
+
+    PyThread_acquire_lock(mutex, WAIT_LOCK);
+    /*
+     * If the lock is still allocated it means that _channelitem_wait_lock_wait
+     * has not timed out.
+     */
+    if (wait_lock->lock_is_allocated == 1) {
+        PyThread_release_lock(wait_lock->lock);
+        wait_lock->is_recv = received;
+
+    } else {
+        free(wait_lock);
+        wait_lock = NULL;
+    }
+    PyThread_release_lock(mutex);
+
+    if (wait_lock == NULL) {
+        PyThread_free_lock(mutex);
+    }
+}
+
+/**
+ * Will wait until the channel item with the wait_lock has been revieved or
+ * for timeout microseconds.
+ * This function must be called by the interpeter sending the message.
+ *
+ * If timeout is:
+ *  - < 0:  then wait until the channelitem has been received.
+ *  - >= 0: then wait until the channelitem has been received or timeout
+ *          microseconds, whichever comes first.
+ *
+ * If channel item is received before this function has timed out then it is
+ * responsible for freeing the wait lock.
+ *
+ * This function returns:
+ * - PY_LOCK_ACQUIRED: The channel item was received by an iterpreter before
+ *                     this function timed out.
+ * - PY_LOCK_FAILURE: The channel item was not received by an interpreter
+ *                    before this function timed out.
+ *
+ * NOTE: This must be called after _channelitem_wait_lock_sent has been called.
+ */
+static PyLockStatus
+_channelitem_wait_lock_wait(_channelitem_wait_lock *wait_lock, int timeout)
+{
+    assert(wait_lock != NULL);
+    assert(wait_lock->is_sent == 1);
+    PyLockStatus lock_rc;
+    Py_BEGIN_ALLOW_THREADS
+    lock_rc = PyThread_acquire_lock_timed(wait_lock->lock, timeout, 0);
+    Py_END_ALLOW_THREADS
+
+    if (lock_rc == PY_LOCK_INTR) {
+        lock_rc = PY_LOCK_FAILURE;
+    }
+
+    PyThread_type_lock mutex = wait_lock->mutex;
+    PyThread_acquire_lock(mutex, WAIT_LOCK);
+    PyThread_free_lock(wait_lock->lock);
```

Comment:
As noted above, I recommend lumping all de-allocation in a single function to mirror `_channelitem_wait_lock_new()` (e.g. `_channelitem_wait_lock_free()`).  Once you have such a function, the best place to call it is in the same function where `_channelitem_wait_lock_new()` was called.